### ISSUE TITLE
Disallow script that inherit from a type be assigned to something else

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#615]
**Describe the project you are working on:**
spaceship project

**Describe the problem or limitation you are having in your project:**
I can't remeber what type a script is without opening and looking at what it inherits 
I keep getting crashes due to accidentally accessing something that doesn't exist

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
basically nodes would only accept a script that inherit from them or a child node.
this would solve some crashes and would overall improve workflow

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
for example this could be assigned to a TextureRect but not a Node...
```
extends TextureRect
```
however this could be assigned to a TextureRect and a Node...
```
extends Node
```

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
this would be used whenever assigning scripts to nodes so very often and I don't know if this can be worked around using a script

**Is there a reason why this should be core and not an add-on in the asset library?:**
this should be added because who would want to assign a script to something it doesn't derive from?.

**Comment by Shadowblitz16** [comment]
this would be made obsolete by [this](https://github.com/godotengine/godot-proposals/issues/681)

**Comment by Zireael07** [comment]
Your proposal is much faster to implement, though.

**Comment by Shadowblitz16** [comment]
@Zireael07 this one would be faster and easier to implement but the one I linked uses script OPP for custom nodes there since nodes and scenes would basically be script

**Comment by Shadowblitz16** [comment]
I'm reopening this due to the unlikely hood of the script and node combination being added.
it would also be nice to have in 4.0.

**Comment by DriNeo** [comment]
It might be cool to simply remove the first "extends" statement and get a dynamic inheritance based on the attached node.

**Comment by pycbouh** [comment]
@DriNeo Scripts can be attached to multiple nodes of different types, so it would make any type assumptions impossible. Also, no `extends` statement already has a special meaning: it automatically extends `Reference`.

**Comment by DriNeo** [comment]
Attaching a script to a node would be exactly like adding the "extends" statement on top. Errors can be raised as usual.

**Comment by pycbouh** [comment]
@DriNeo What I'm saying is that the same script can be attached to any number of nodes, including nodes of different types. Therefore making any assumptions about scripts base type would be impossible. Scripts don't know to which nodes they are attached.

And like I've said, having no `extends` specified at the top automatically extends `Reference`, so what you propose conflicts with existing behavior unrelated to nodes.

**Comment by Shadowblitz16** [comment]
@DriNeo also specifying what a script extends allows static typing.
so you can make sure your script has the functionality it needs to run

**Comment by KoBeWi** [comment]
Would https://github.com/godotengine/godot/pull/81886 resolve this? It does not block attaching the script, but it's a more feasible solution.

**Comment by Shadowblitz16** [comment]
> Would [godotengine/godot#81886](https://github.com/godotengine/godot/pull/81886) resolve this? It does not block attaching the script, but it's a more feasible solution.

No. if scriptA inherits from Node2D it should not be able to be attached to a Node3D.
